### PR TITLE
:sparkles: Add Monte-Carlo Dropout

### LINKED ISSUE #40: :sparkles: Add an MCDropout Baseline
I have added dropout to standard ResNet in [de31452](https://github.com/ENSTA-U2IS/torch-uncertainty/commit/de314523bf9fc6ca6c06e4a4f3046097ac45446a). It could be interesting to add an MCDropout Ensemble baseline. To do this, the following steps could be considered:

- [ ] Create a new baseline in torch-uncertainty/torch_uncertainty/baselines/classification/resnet.py in cls.ensemble
- [ ] Find a solution to apply dropout in eval mode (potentially change nn.Dropout to F.dropout and replace self.training by true when the version is MCdropout)
- [ ] Add tests
- [ ] Add a tutorial to explain how to use the baseline
- [ ] Add the reference to the documentation
- [ ] Add a parameter to enable last-layer dropout

### PR DESCRIPTION
**Pull request by badrmarani**
This PR fixes #40. The main changes are:
- [x] Create a new baseline.
- [x] Find a solution to apply dropout in `eval` mode.
- [x] Add tests.
- [x] Add a tutorial to explain how to use the baseline.
- [x] Add the reference to the documentation.
- [x] Add the parameter `enable_last_layer_dropout` to enable last-layer dropout.
- [x] Extend this method to VGG and WideResNet.

### CODE REVIEW

**Comment by o-laurent** [comment]
Hi! Thanks for the great work, again! 

I'll have a more thorough look at it next week. I've skimmed your PR and am just wondering how we could adapt your baseline to the different variants of MC Dropout. It may be a question worth considering, given your interest in its variants!

Then, there is also the question of the number of inference passes. I agree that it is more optimal to avoid the for loop and increase the batch size, but it may lead, in some cases, to increased VRAM needs (for instance, with 1000 passes?). We could let the user select the test batch size if wanted (that could be reduced to 1, for instance, if we want to compute the mean on many passes). It seems to be the simplest solution, but it is not completely safe & it adds a new parameter to the CLI.

File: auto_tutorials_source/tutorial_mc_dropout.py [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+    args = init_args(network=ResNet, datamodule=MNISTDataModule)
+
+net_name = "mc-dropout-resnet18-mnist"
+
+# datamodule
+args.root = str(root / "data")
+dm = MNISTDataModule(**vars(args))
+
+
+model = resnet18(
```
Comment (o-laurent) [inline auto_tutorials_source/tutorial_mc_dropout.py:89]: We may switch to a lenet to avoid too-long computations as we don't have GitHub runners yet.

File: torch_uncertainty/models/resnet/std.py [hunk 2]
```diff
@@ -197,11 +199,15 @@ def __init__(
         dropout_rate: float,
         groups: int,
         style: str = "imagenet",
+        num_estimators: int = None,
+        enable_last_layer_dropout: bool = False,
```
Comment (o-laurent) [inline torch_uncertainty/models/resnet/std.py:203]: I'll change this variable to last__layer_dropout or ll_dropout. @alafage, the latter is not explicit enough?

File: torch_uncertainty/models/resnet/std.py [hunk 3]
```diff
@@ -197,11 +199,15 @@ def __init__(
         dropout_rate: float,
         groups: int,
         style: str = "imagenet",
+        num_estimators: int = None,
```
Comment (o-laurent) [inline torch_uncertainty/models/resnet/std.py:202]: I agree it's better to keep num_estimators as a parameter name to avoid increasing their number, but it may sound strange.

File: torch_uncertainty/models/wideresnet/std.py [hunk 4]
```diff
@@ -186,7 +198,10 @@ def wideresnet28x10(
     in_channels: int,
     num_classes: int,
     groups: int = 1,
+    dropout_rate: float = 0.0,
```
Comment (o-laurent) [inline torch_uncertainty/models/wideresnet/std.py:201]: I'd leave 0.3 as the default value since WideResNet is intended to be trained with dropout if I recall the paper correctly.

**Review by o-laurent — COMMENTED** [review]
Thanks for the PR. I will make some modifications, but having it in the library is great. If you want, you can join the Discord server whose link I added in the readme of the dev branch!

File: torch_uncertainty/models/resnet/std.py (hunk 3 above)
Comment (o-laurent) [reply torch_uncertainty/models/resnet/std.py:202]: I've added a docstring to make it a bit more explicit.

**Comment by o-laurent** [comment]
Merging after discussions with @alafage.

**Comment by o-laurent** [comment]
I had to make a few changes to avoid errors. Merging now.
